Join NDEFReader and NDEFWriter

zolkis opened the pull request:
Re-applying #568.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/zolkis/web-nfc/pull/601.html" title="Last updated on Nov 19, 2020, 7:08 PM UTC (e82d4b2)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/web-nfc/601/934c053...zolkis:e82d4b2.html" title="Last updated on Nov 19, 2020, 7:08 PM UTC (e82d4b2)">Diff</a>

Commit:
Merge NDEFReader and NDEFWriter

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

beaufortfrancois commented:
Quick question: Is it clear for web developers that `onerror` will get fired only when scan has started.
I can see why some web developers would think this code would work.

```js
const ndef = new NDEFReader();
ndef.onerror = event => {
  /* Was there an error while writing? */
}
ndef.write('hello world');
```

zolkis commented:
Good question, we need to clarify that with the two objects merged.

zolkis commented:
So should we rename `onerror` to `onreaderror`, or get rid of it alltogether and rely on the Promise, like with write()?

Commit:
Rename onerror to onreaderror. Replace activated reader list with pending read tuple.

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

beaufortfrancois commented on index.html:
```diff
@@ -1980,30 +1974,26 @@ <h2>The <dfn>record type</dfn> string</h2>
       constructor();
 
-      attribute EventHandler onerror;
+      attribute EventHandler onreadingerror;
```
Nit: Move `onreadingerror` below `onreading`

zolkis replied:
done

beaufortfrancois commented on index.html:
```diff
@@ -2041,10 +2031,11 @@ <h2>The <dfn>record type</dfn> string</h2>
   </p>
   <p>
-    The <dfn>NDEFWriter</dfn> is an object used for writing data to an <a>NFC tag</a>.
+    The <dfn>NDEFReader</dfn> is an object used for reading or writing data when
```
I believe this is redundant with the first paragraph of "The NDEFReader object"

zolkis replied:
Removed.

beaufortfrancois commented on index.html:
```diff
@@ -1980,30 +1974,26 @@ <h2>The <dfn>record type</dfn> string</h2>
+<section> <h2>The NDEFReader object</h2>
+  <p>
+    The <dfn>NDEFReader</dfn> is an object that exposes NFC functionality to the
+    <a>browsing context</a>: reading data when a device, such as a tag, is within
```
I believe we should explicitly mention NDEF Message for reading like we do for writing

zolkis replied:
Done.

beaufortfrancois commented on index.html:
```diff
@@ -981,12 +980,10 @@ <h4>
       Detecting if Web NFC is supported can be done by checking NDEFReader
-      and/or NDEFWriter objects. Note that this does not guarantee that NFC
-      hardware is available.
+      objects. Note that this does not guarantee that NFC hardware is available.
```
Nit: `object`, not `objects`.

zolkis replied:
Fixed.

beaufortfrancois commented on index.html:
```diff
@@ -981,12 +980,10 @@ <h4>
     <pre class="example">
-      if ('NDEFReader' in window) { /* ... Scan NDEF Tags */ }
-      if ('NDEFWriter' in window) { /* ... Write NDEF Tags */ }
+      if ('NDEFReader' in window) { /* ... Scan or write NDEF Tags */ }
```
Nit: `Scan and...`

zolkis replied:
Fixed.

beaufortfrancois reviewed (commented):
Would you mind also updating the the EXPLAINER.md file and other files where NDEFWriter is referenced?

Commit:
Update explainer because NDEFReader/NDEFWriter merge

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

Commit:
Update TypeScript definitions because NDEFReader/NDEFWriter merge

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

zolkis commented:
Amended the TypeScript definition to also remove `ignoreRead` and scan options no longer supported.

arskama commented:
> Quick question: Is it clear for web developers that `onerror` will get fired only when scan has started.
> I can see why some web developers would think this code would work.
> 
> ```js
> const ndef = new NDEFReader();
> ndef.onerror = event => {
>   /* Was there an error while writing? */
> }
> ndef.write('hello world');
> ```

@beaufortfrancois 
I would like to have a clarification on your comment (it s me not understanding :) )

On the chromium implementation, there are 2 places where onerror is fired.
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/nfc/ndef_reader.cc;l=217;drc=126a458910eb392beb0c56601c89233cd82d80b5

and

https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/nfc/nfc_proxy.cc;l=122;drc=126a458910eb392beb0c56601c89233cd82d80b5

which are 2 different cases, right?
So I think we need two different handling for those, since the first one is not really a readingerror?

zolkis commented:
Since we merged NDEFReader (which had an `onerror` event) and NDEFWriter (which did not have it), someone may assume that in the joined NDEFReader `onerror` is also relevant to writing, which is not the case. Therefore we renamed `onerror` to `onreadingerror`.

In the impl `onerror` conveys the NFC read errors (either due to access problems, or technical faults) to the readers.

Looks like since filters are gone, we have only one reader, and `NFCProxy` could be simplified (`readers_` removed etc), in conformance with the new internal slots (`PendingReader` instead of `ActivatedReaderList`) of `NDEFReader`.

beaufortfrancois commented:
Argh I see now... You're right @arskama, the first one "NDEFReader::OnMojoConnectionError()" is not per-say a reading error but an internal communication issue that can happen in Chromium.
If I remember correctly though, there's no way to recover from this.

The old behaviour still triggered an error even though it was NOT specifically a reading error.

With that in mind, maybe keeping `onerror` makes sense.
For info, this will also get triggered if an internal error happens during a write NFC operation and `write()` promise will reject as well...

Let's keep `onerror`.
What do you folks think?

zolkis commented:
> Let's keep onerror. What do you folks think?

So that error could have happened with the write(), too.
When does that error come up typically? Can we handle that via the scan() and write() promise?

zolkis commented:
>  "NDEFReader::OnMojoConnectionError()" is not per-say a reading error but an internal communication issue that can happen in Chromium.

Do I read that correctly: it seems to reject the promise with that error and _also_ fire an `onerror`? Is the latter needed?
Edit: it's covering the case the error comes _after_ the promise is settled. But does that actually (typically) happen?

beaufortfrancois commented:
> > Let's keep onerror. What do you folks think?
> 
> So that error could have happened with the write(), too.
> When does that error come up typically? Can we handle that via the scan() and write() promise?

scan will succeed when scan has started successfully, but this error could happen way after.
re: write, it won't be an issue as if this error happens, the promise will reject right away, and the onerror event will also get fired.

> > "NDEFReader::OnMojoConnectionError()" is not per-say a reading error but an internal communication issue that can happen in Chromium.
> 
> Do I read that correctly: it seems to reject the promise with that error and _also_ fire an `onerror`? Is the latter needed?
> Edit: it's covering the case the error comes _after_ the promise is settled. But does that actually (typically) happen?

it is an internal issue which "shouldn't" happen but can happen.

zolkis commented:
OK, I will add a commit the revert to `onerror` and explain the behavior for reading and writing.

zolkis commented:
Question, @beaufortfrancois @kenchris 

With filters gone, we can still create multiple NDEFReader instances, each of which has a scan() method that operate on the same NFC HW and have an `onreading` event that can take callbacks.

How should we handle this? 
1. Allow all the NDEFReader object instances, but keep a single reader in the internal slot and add the event listeners to its event.
2. Allow all the object instance, and keep all of these in the `activated reader list`. In the impl, there is a single NFC proxy, though.
3. Allow a single instance of NDEFReader in the given context and decide how to expose it: provide a getter/factory method for it, or let it be a static object, e.g. `navigator.nfc` (we used to be there).

Commit:
Add navigator.nfc and navigator.nfc.ndef objects

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

zolkis commented:
Sorry, I needed to amend since I forgot to update the explainer and the TS definitions.
So, as agreed, now there is only one NDEFReader object that is accessed via `navigator.nfc.ndef`.
NFC state could be also moved to the `navigator.nfc` object, but for the moment it looked good on the settings object.
However, I moved all relevant content (NFC state, handling NFC adapters, obtaining permission, handling visibility change. releasing NFC) under `navigator.nfc`.

Commit:
Use navigator.nfc.ndef directly in examples

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

beaufortfrancois commented on EXPLAINER.md:
````diff
@@ -92,15 +92,15 @@ when required, about granting access to Web NFC. It means sending and receiving
 ```js
-const reader = new NDEFReader();
- 
+const reader = navigator.nfc.ndef;
````
I think it's fair to use `navigator.nfc.ndef` inline.

```js

async function startScanning() {
  await navigator.nfc.ndef.scan();
  navigator.nfc.ndef.onreading = event => {
    /* handle NDEF messages */
  };
}
```

beaufortfrancois commented on EXPLAINER.md:
````diff
@@ -123,10 +123,10 @@ USB](https://wicg.github.io/webusb/) in that it offers secure low level access
 
 ```js
-const reader = new NDEFReader();
+const reader = navigator.nfc.ndef;
````
Same here.
```js
await navigator.nfc.ndef.scan({ recordType: "example.com:smart-poster" });
navigator.nfc.ndef.onreading = event => {
  const externalRecord = event.message.records.find(
    record => record.type == "example.com:smart-poster"
  );

  let action, text;

  for (const record of externalRecord.toRecords()) {
    if (record.recordType == "text") {
      const decoder = new TextDecoder(record.encoding);
      text = decoder.decode(record.data);
    } else if (record.recordType == "act") {
      action = record.data.getUint8(0);
    }
  }

  console.log({text, action});
};
```

beaufortfrancois commented on EXPLAINER.md:
```diff
@@ -123,10 +123,10 @@ USB](https://wicg.github.io/webusb/) in that it offers secure low level access
-const reader = new NDEFReader();
+const reader = navigator.nfc.ndef;
 
 await reader.scan({ recordType: "example.com:smart-poster" });
```
`recordType` has been deprecated. Let's use `  await navigator.nfc.ndef.scan();`

beaufortfrancois commented on EXPLAINER.md:
```diff
@@ -162,13 +162,11 @@ const abortController = new AbortController();
-await writer.push("foo", { signal: abortController.signal });
- 
+
+await navigator.nfc.ndef.scan({ signal: abortController.signal });
```
I'd remove those two `await` otherwise `document.querySelector("#abortButton").onclick` declaration wouldn't be valid.

beaufortfrancois commented on index.html:
```diff
@@ -980,13 +979,14 @@ <h4>
     <pre class="example">
-      if ('NDEFReader' in window) { /* ... Scan NDEF Tags */ }
-      if ('NDEFWriter' in window) { /* ... Write NDEF Tags */ }
+      if (navigator.nfc && navigator.nfc.ndef) {
```
```suggestion
      if (navigator.nfc?.ndef) {
```

beaufortfrancois commented on index.html:
```diff
@@ -1043,9 +1042,8 @@ <h4>
-      const ndef = new NDEFReader();
       try {
-        ndef.write({
+        navigator.nfc.ndef.write({
```
Use `await` to catch exception.
```suggestion
        await navigator.nfc.ndef.write({
```

beaufortfrancois commented on index.html:
```diff
@@ -1082,14 +1079,15 @@ <h4>
         return new Promise((resolve, reject) => {
-          reader.addEventListener("reading", event => {
+          navigator.nfc.ndef.addEventListener("reading", event => {
             // Check if we want to write to this tag, or reject.
```
Where do we check here?

zolkis replied:
The example & comment was by @kenchris

beaufortfrancois commented on index.html:
```diff
@@ -1106,8 +1104,8 @@ <h4>
     <pre class="example">
-      const reader = new NDEFReader();
-      reader.onreading(event => console.log("We read a tag!"));
+      const ndef = navigator.nfc.ndef;
```
May we try without the shortcut `const ndef` for consistency and use `navigator.nfc.ndef` everywhere?

beaufortfrancois commented on index.html:
```diff
@@ -1167,17 +1165,16 @@ <h4>
           ctlr.signal.onabort = reject;
-          reader.addEventListener("reading", event => {
+          navigator.nfc.ndef.addEventListener("reading", event => {
             ctlr.abort();
```
I'm not sure about this. How does this work exactly?

beaufortfrancois replied:
@kenchris Do we need `catch(err => reject(err));`?

beaufortfrancois replied:
Let's keep it like this.

kenchris replied:
I think we need to remove the ctlr.signal.onabort because you don't need to reject if it actually succeeds getting a reading.

kenchris replied:
It works the way that if you get an event it resolves with the data and aborts the reader (stops reading).

if it fails for some reason, it will just reject

kenchris replied:
it might actually be working today with ctlr.signal.onabort = reject; because abort is running async code and only one of reject/resolve can be called (the next is ignored)

kenchris replied:
Maybe we should explain how it works :)

beaufortfrancois replied:
Please open a new PR for this ;)

beaufortfrancois commented on index.html:
```diff
@@ -2139,25 +2012,26 @@ <h2>The <dfn>record type</dfn> string</h2>
       <td>
-        A <a>set</a> of {{NDEFReader}} instances.
+        A &lt;|promise:Promise|, |reader:NDEFReader|&gt; tuple where |promise|
+        holds a {{Promise}} and |reader| holds an {{NDEFReader}}.
```
Why not `pending` as in "PendingWrite" description?

zolkis replied:
fixed

beaufortfrancois commented on index.html:
```diff
@@ -1979,139 +1962,29 @@ <h2>The <dfn>record type</dfn> string</h2>
-    accepted by the {{NDEFWriter/write()}} method.
+  <p>The
+    <dfn>ndef</dfn> property is an {{NDEFReader}} object that provides
+    <a>NDEF</a> tag read and write functionality.
```
Shall we use NFC tags consistently?

zolkis replied:
True, changed. I was thinking not all NFC tags were NDEF, so using "NDEF tag" was more appropriate for NDEFReader.

beaufortfrancois commented on web-nfc.d.ts:
```diff
@@ -6,6 +6,12 @@
 // https://w3c.github.io/web-nfc/#actual-idl-index
 
+interface Navigator {
+  nfc: NFC
```
```suggestion
  readonly nfc: NFC
```

beaufortfrancois commented on web-nfc.d.ts:
```diff
@@ -6,6 +6,12 @@
+  nfc: NFC
+}
+declare class NFC {
+  ndef: NDEFReader
```
```suggestion
  readonly ndef: NDEFReader
```

beaufortfrancois reviewed (commented):
Here's my first feedback to get your thoughts.

beaufortfrancois commented:
What does happen with code below?
```js
const promise1 = navigator.nfc.ndef.scan();
const promise2 = navigator.nfc.ndef.scan();
```
Which `promise1` of `promise2` reject?
It is not clear to me.

Same for `write()`...

beaufortfrancois reviewed (commented):
We shall update `Separate objects vs navigator.nfc` and `Separate objects for reader/writer` sections in EXPLAINER.md

zolkis commented:
Summarized the discussion about constructors vs navigator.nfc in #602.
Also, some options to address remaining issues like the ones @beaufortfrancois mentioned above.

zolkis commented:
OK, now we have `navigator.nfc` that contains `ndef` of type `NDEFReader` (exposing scan(), write(), onreading, onreadingerror).

I will make a branch from here to explore option 2 from #602 [comment](https://github.com/w3c/web-nfc/issues/602#issuecomment-707945243). That is, `navigator.nfc` would contain `ontagfound`, `onerror`,  and `ndef` of type `NDEFReader` that contains `read()` and `write()`. 
Scanning (NFC polling) is separated from reads, `ignoreRead` is not needed any more since there is no automatic content read from tags, and internal slots for pending read and write are not needed, either. Reading and dispatching algorithms become much simpler. Also, this would be closer to the Android NFC API usage flow, and more future proof.

This would permit later improvements, such as the `ontagfound` event could contain NFC technologies supported, but since that is restricted to NDEF at the moment, we can use a simple event. However, we could use flags already available from Android, such as `writeable`, `formattable` etc.
Future NFC tech could get their API namespace under `navigator.nfc`, for instance `navigator.nfc.isodep` exposing `transceive()` etc. The same `navigator.nfc.ontagfound` could be used (with expanded content, so backwards compatible).
Also, a `navigator.nfc.ontagreleased` event could be added later, if a developer need arises (it's quite common in other APIs, but will only add here when there are clear use cases for it).

The minimal delta from this point that supports the current functionality would be quite small, but let's see how

Commit:
Revert "Add navigator.nfc and navigator.nfc.ndef objects"

zolkis commented:
Reverted attaching to `navigator`. That version is saved in PR #605

beaufortfrancois commented on EXPLAINER.md:
```diff
@@ -162,13 +162,12 @@ const abortController = new AbortController();
+const ndef = new NDEFReader();
+await ndef.scan({ signal: abortController.signal });
+
+await ndef.push("foo", { signal: abortController.signal });
```
It should  be `write` not `push`

beaufortfrancois commented on EXPLAINER.md:
````diff
@@ -93,14 +93,14 @@ info when users tap NFC devices can be done smoothly once permission is granted.
 
 ```js
 const reader = new NDEFReader();
````
Use `const ndef` here and below.

zolkis replied:
Hmm. Looks like these changes have been lost in the revert, need to re-apply.

beaufortfrancois commented on EXPLAINER.md:
```diff
@@ -93,14 +93,14 @@ info when users tap NFC devices can be done smoothly once permission is granted.
- 
+
 const nfcPermissionStatus = await navigator.permissions.query({ name: "nfc" });
 if (permissionStatus.state === "granted") {
```
Use `nfcPermissionStatus`

beaufortfrancois commented on EXPLAINER.md:
````diff
@@ -123,7 +123,7 @@ USB](https://wicg.github.io/webusb/) in that it offers secure low level access
+implementations.
 
 ```js
 const reader = new NDEFReader();
````
Use `ndef` instead of `reader`.

beaufortfrancois commented on EXPLAINER.md:
```diff
@@ -234,7 +233,7 @@ The reader and writer objects could have been merged into one single object, but
 separate objects.
- 
+
 ## Considered alternatives
```
Please remove `### Separate objects for reader/writer` section above.

zolkis replied:
Yes.

zolkis replied:
Wait, that's only a title for a problem (explaining the API choice). So what should be the new name?

zolkis replied:
Updated

beaufortfrancois reviewed (changes requested):
Here's my feedback for the explainer.
I'll now review spec changes.

beaufortfrancois commented on index.html:
```diff
@@ -1068,10 +1065,10 @@ <h4>
       let ignoreRead = false;
 
-      reader.onreading(event => {
+      ndef.onreading(event => {
```
This is not valid in JS. Please do not forget to remove extra `)` at the end of this function.

```suggestion
      ndef.onreading = (event) => {

zolkis replied:
fixed, thanks

beaufortfrancois replied:
It is still not fixed.

beaufortfrancois commented on index.html:
```diff
@@ -1106,8 +1103,8 @@ <h4>
-      const reader = new NDEFReader();
-      reader.onreading(event => console.log("We read a tag!"));
+      const ndef = new NDEFReader();
+      ndef.onreading(event => console.log("We read a tag!"));
```
```suggestion
      ndef.onreading = (event) => console.log("We read a tag!");

zolkis replied:
Done

beaufortfrancois commented on index.html:
```diff
@@ -2056,12 +2043,20 @@ <h2>The <dfn>record type</dfn> string</h2>
+        <td><dfn>[[\ScanAbortSignal]]</dfn></td>
+        <td>`undefined`</td>
+        <td>
+          The {{NDEFScanOptions}}.signal to abort the operation.
```
Shall we specify it's only about the scan operation?

zolkis replied:
Done

beaufortfrancois commented on index.html:
```diff
@@ -3478,27 +3446,25 @@ <h3><dfn>Writing content</dfn></h3>
               <li>
                 If |key| equals "`signal`" and |value| is not `undefined`, set
-                |reader|.<a>[[\Signal]]</a> to |value|.
+                |reader|.<a>[[\ScanAbortSignal]]</a> to |value|.
```
How about doing the same as NDEFWriter and not use ScanAbortSignal, simply the signal from the dictionary member?

![image](https://user-images.githubusercontent.com/634478/99400706-92c6f180-28e7-11eb-94d7-1e2a3c9f212f.png)

zolkis replied:
Should we then also remove the internal slot? (my pleasure)

beaufortfrancois reviewed (changes requested):
See my comments.

Commit:
Update the explainer

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

beaufortfrancois commented on index.html:
```diff
@@ -1106,8 +1103,8 @@ <h4>
-      const reader = new NDEFReader();
-      reader.onreading(event => console.log("We read a tag!"));
+      const ndef = new NDEFReader();
+      ndef.onreading = event => console.log("We read a tag!"));
```
```suggestion
      ndef.onreading = event => console.log("We read a tag!");
```

beaufortfrancois reviewed (commented):
Re: https://github.com/w3c/web-nfc/pull/601#discussion_r525211173 We could remove it if we decide to remove NDEFWriteOptions as well from the slots.

Commit:
Update examples

Co-authored-by: François Beaufort <beaufort.francois@gmail.com>

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

zolkis commented:
>  We could remove it if we decide to remove NDEFWriteOptions as well from the slots.

done

beaufortfrancois commented:
> > We could remove it if we decide to remove NDEFWriteOptions as well from the slots.
> 
> done

Shall we remove WriteMessage as well then?
![image](https://user-images.githubusercontent.com/634478/99408926-12a58980-28f1-11eb-9cf7-e56eb12dc6d7.png)

beaufortfrancois commented:
> OK, I will add a commit the revert to `onerror` and explain the behavior for reading and writing.

FYI I'm still seeing `readingerror` in latest patch

zolkis commented:
> BTW the linking is broken somewhere, as it is actually used in the "start the NFC write" algorithm.

Thanks for catching!

beaufortfrancois commented:
@zolkis To fix linking use `<a>[[\WriteOptions]]</a>` instead of `[[\WriteOptions]]` only. Same for `WriteMessage`.

Commit:
Fix internal slots linking

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

zolkis commented:
> To fix linking use <a>[[\WriteOptions]]</a> instead of [[\WriteOptions]] only. Same for WriteMessage.

Thanks, they are fixed now. Also checked the other internal slots.

beaufortfrancois commented on web-nfc.d.ts:
```diff
@@ -43,25 +43,18 @@ declare interface NDEFRecordInit {
   constructor()
-  onerror: (error: any) => void
   onreading: (event: NDEFReadingEvent) => void
+  onreadingerror: (error: any) => void
```
`onreadingerror` doesn't expose error or anything actually.

zolkis replied:
should be `onreadingerror: (event: Event) => void` ?

beaufortfrancois reviewed (approved):
With few nits, this change looks good to me.

beaufortfrancois committed:
Remove origin trial section

beaufortfrancois committed:
Update ReSpec header with new "group" option

Commit:
Fix onreadingerror TS def

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

beaufortfrancois commented on web-nfc.d.ts:
```diff
@@ -43,25 +43,18 @@ declare interface NDEFRecordInit {
   constructor()
-  onerror: (error: any) => void
   onreading: (event: NDEFReadingEvent) => void
+  onreadingerror: (error: Event) => void
```
We should ask TS experts but I'd go with what others do.

See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/w3c-web-usb/index.d.ts#L110-L112 and https://github.com/DefinitelyTyped/DefinitelyTyped/blob/927064b9a3a5f1075f3560eefc923860be7e0b53/types/web-bluetooth/index.d.ts#L163

```
declare class NDEFReader extends EventTarget {
  constructor()
  onreading: (this: this, ev: NDEFReadingEvent) => any
  onreadingerror: (this: this, ev: Event) => any
```

zolkis replied:
right, https://www.valentinog.com/blog/this/

zolkis replied:
amended the TS commit
